### PR TITLE
{bio}[foss/2020b] ABRA2 v2.23, CRISPR-DAV v2.3.4, Excel-Writer-XLSX v1.09, FLASH v2.2.00, PRINSEQ v0.20.4, pysamstats v1.1.2  w/ Python 3.8.6

### PR DESCRIPTION
**Pull request by deniskristak**
(created using `eb --new-pr`)

requires ~https://github.com/easybuilders/easybuild-easyblocks/pull/2487~ (CRISPR-DAV easyblock) + ~#13762~ (`naturalsort` extension for `R`)

### CODE REVIEW

File: easybuild/easyconfigs/e/Excel-Writer-XLSX/Excel-Writer-XLSX-1.09-foss-2020b.eb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+dependencies = [
+    ('Perl', '5.32.0'),
+    ('Archive-Zip', '1.68'),
+]
+
+options = {'modulename': 'Excel::Writer::XLSX'}
+
+sanity_check_paths = {
+    'files': ['bin/extract_vba'],
```
Comment (boegel) [inline easybuild/easyconfigs/e/Excel-Writer-XLSX/Excel-Writer-XLSX-1.09-foss-2020b.eb:25]: Does this command support `--help` or `--version`?

If so, please add that as a sanity check command:

```python
sanity_check_commands = ["extract_vba --help"]

File: easybuild/easyconfigs/n/naturalsort/naturalsort-0.1.3-foss-2020b.eb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+name = 'naturalsort'
+version = '0.1.3'
+
+homepage = 'https://github.com/kos59125/naturalsort/'
+description = """The package provides functions related to human natural ordering."""
+
+toolchain = {'name': 'foss', 'version': '2020b'}
+
+source_urls = ['https://cran.r-project.org/src/contrib/']
+sources = ['%s_%s.tar.gz' % (name, version)]
```
Comment (boegel) [inline easybuild/easyconfigs/n/naturalsort/naturalsort-0.1.3-foss-2020b.eb:12]: Please use templates:

```python
sources = ['%(name)ss_%(version)ss.tar.gz']

File: easybuild/easyconfigs/p/pysamstats/pysamstats-1.1.2-foss-2020b.eb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    ('Python', '3.8.6'),
+    ('Pysam', '0.16.0.1'),
+    ('SciPy-bundle', '2020.11'),
+    ('PyTables', '3.6.1'),
+]
+
+sanity_pip_check = True
+sanity_check_paths = {
+    'files': [],
+    'dirs': ['lib/python3.8/site-packages'],
```
Comment (boegel) [inline easybuild/easyconfigs/p/pysamstats/pysamstats-1.1.2-foss-2020b.eb:30]: This is the default sanity check, so you can remove it

File: easybuild/easyconfigs/p/pysamstats/pysamstats-1.1.2-foss-2020b.eb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+toolchain = {'name': 'foss', 'version': '2020b'}
+
+sources = [SOURCE_TAR_GZ]
+checksums = ['6d8b41934fdc700552d689f473860c64525a88600460344d0cac2095424750c6']
+
+use_pip = True
+download_dep_fail = True
+
+preinstallopts = "sed -i -e 's/ (<0.16)//g' setup.py && "
```
Comment (boegel) [inline easybuild/easyconfigs/p/pysamstats/pysamstats-1.1.2-foss-2020b.eb:18]: Add a comment above this line to explain it?

**Review by boegel — CHANGES_REQUESTED** [review]

**Comment by deniskristak** [comment]
please dont merge yet, ive added crispr-dav, but it still needs work

**Comment by deniskristak** [comment]
must be used with https://github.com/easybuilders/easybuild-easyblocks/pull/2487

File: easybuild/easyconfigs/c/crispr-dav/crispr-dav-2.3.4-foss2020b_remove_hardcoding.patch [hunk 5]
```diff
@@ -0,0 +1,128 @@
+# changing example command to contain absolute paths + tabs escaping + changing plotting type to cairo
```
Comment (boegel) [inline easybuild/easyconfigs/c/crispr-dav/crispr-dav-2.3.4-foss2020b_remove_hardcoding.patch:1]: @deniskristak Please mention yourself as author?

**Review by boegel — CHANGES_REQUESTED** [review]

**Comment by deniskristak** [comment]
pushed to easyblock PR as well

**Comment by deniskristak** [comment]
ready for review, @boegel

File: easybuild/easyconfigs/c/CRISPR-DAV/CRISPR-DAV-2.3.4-foss-2020b.eb [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    '41475a09754d65cc2c88d4161be6eacdfc98242bccbe49778c4f1f74210a6cb2',
+]
+
+dependencies = [
+    ('R', '4.0.3'),
+    ('Perl', '5.32.0'),
+    ('NGS', '2.10.9'),
+    ('Pysam', '0.16.0.1'),
+    ('Excel-Writer-XLSX', '1.09'),
+    ('naturalsort', '0.1.3'),
```
Comment (boegel) [inline easybuild/easyconfigs/c/CRISPR-DAV/CRISPR-DAV-2.3.4-foss-2020b.eb:25]: I think it makes more sense to install `naturalsort` as an additional extension in the `R` module, since it's a standard `R` package, see #13762

File: easybuild/easyconfigs/a/ABRA2/ABRA2-2.23-GCC-10.2.0.eb [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+postinstallcmds = ["cd %(installdir)s && mv abra2-%(version)s-jar-with-dependencies.jar abra2-%(version)s.jar"]
+
+sanity_check_paths = {
+    'files': ['abra2-%(version)s.jar', 'bin/abra', 'lib/libAbra.%s' % SHLIB_EXT],
+    'dirs': [],
+}
+
+modextravars = {
+    "LC_ALL": "en_US.UTF-8",
```
Comment (boegel) [inline easybuild/easyconfigs/a/ABRA2/ABRA2-2.23-GCC-10.2.0.eb:45]: @deniskristak Hmm, why is this needed? That looks a bit fishy, it doesn't belong in a module generated by EasyBuild...

File: easybuild/easyconfigs/c/CRISPR-DAV/CRISPR-DAV-2.3.4-foss-2020b.eb [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+homepage = 'https://github.com/pinetree1/crispr-dav/'
+description = """CRISPR-DAV is a pipeline to analyze
+amplicon-based NGS data of CRISPR clones in a high throughput manner."""
+
+toolchain = {'name': 'foss', 'version': '2020b'}
+
+source_urls = ['https://github.com/pinetree1/crispr-dav/archive/']
+sources = ['v%(version)s.tar.gz']
+patches = ['crispr-dav-2.3.4-foss2020b_remove_hardcoding.patch']
```
Comment (boegel) [inline easybuild/easyconfigs/c/CRISPR-DAV/CRISPR-DAV-2.3.4-foss-2020b.eb:12]: The script is not specific to `foss/2020b` at all, so that part should be removed from the patch name.
I would also use `CRISPR-DAV` and `%(version)s`, so:

```python
patches = ['CRISPR-DAV-%(version)s-foss2020b_remove_hardcoding.patch']
```

Don't forget the rename the patch file itself accordingly, and note that there are actually *two* patches with a very similar name currently, one should be removed.

File: easybuild/easyconfigs/f/FLASH/FLASH-2.2.00-foss-2020b.eb [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+toolchain = {'name': 'foss', 'version': '2020b'}
+
+source_urls = ['https://github.com/dstreett/FLASH2/archive/']
+sources = ['%(version)s.zip']
+checksums = ['1e54b2dd7d21ca3e0595a3ffdd27ef3098f88c4de5b9302ec5ea074b49b79960']
+
+files_to_copy = [(['flash2'], 'bin')]
+
+postinstallcmds = ['ln -s %(installdir)s/bin/flash2 %(installdir)s/bin/flash']
```
Comment (boegel) [inline easybuild/easyconfigs/f/FLASH/FLASH-2.2.00-foss-2020b.eb:23]: Please make this a relative symlink:

```python
postinstallcmds = ["cd %(installdir)s/bin && ln -s flash2 flash"]
```

Also, can we add a sanity check command like `flash --help` or whatever?

File: easybuild/easyconfigs/n/naturalsort/naturalsort-0.1.3-foss-2020b.eb [hunk 10]
```diff
@@ -0,0 +1,24 @@
+easyblock = 'RPackage'
+
+name = 'naturalsort'
```
Comment (boegel) [inline easybuild/easyconfigs/n/naturalsort/naturalsort-0.1.3-foss-2020b.eb:3]: This easyconfig file should be removed, since we're adding `naturalsort` as an extension to R, see #13762

File: easybuild/easyconfigs/p/pysamstats/pysamstats-1.1.2-foss-2020b.eb [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
+dependencies = [
+    ('Python', '3.8.6'),
+    ('Pysam', '0.16.0.1'),
+    ('SciPy-bundle', '2020.11'),
+    ('PyTables', '3.6.1'),
+]
+
+sanity_pip_check = True
```
Comment (boegel) [inline easybuild/easyconfigs/p/pysamstats/pysamstats-1.1.2-foss-2020b.eb:28]: add `pysamstats --help` as sanity check command?

File: easybuild/easyconfigs/e/Excel-Writer-XLSX/Excel-Writer-XLSX-1.09-foss-2020b.eb [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+    'dirs': ['lib', 'bin'],
```
Comment (boegel) [inline easybuild/easyconfigs/e/Excel-Writer-XLSX/Excel-Writer-XLSX-1.09-foss-2020b.eb:26]: Slightly better (to make sure that the Perl module stuff doesn't get added to the `Perl` install dir itself somehow):

```python
sanity_check_paths = {
    'files': ['bin/extract_vba'],
    'dirs': ['lib/perl5/site_perl/%(perlver)s/Excel/Writer/XLSX'],
}
```

**Review by boegel — CHANGES_REQUESTED** [review]

File: easybuild/easyconfigs/a/ABRA2/ABRA2-2.23-GCC-10.2.0.eb (hunk 7 above)
Comment (deniskristak) [reply easybuild/easyconfigs/a/ABRA2/ABRA2-2.23-GCC-10.2.0.eb:45]: ABRA2 needs it because of this:
https://github.com/mozack/abra2/issues/25

File: easybuild/easyconfigs/c/CRISPR-DAV/CRISPR-DAV-2.3.4-foss-2020b.eb (hunk 6 above)
Comment (deniskristak) [reply easybuild/easyconfigs/c/CRISPR-DAV/CRISPR-DAV-2.3.4-foss-2020b.eb:25]: so should I remove it altogether from here?
I see it has been added to `R` already

File: easybuild/easyconfigs/n/naturalsort/naturalsort-0.1.3-foss-2020b.eb (hunk 10 above)
Comment (deniskristak) [reply easybuild/easyconfigs/n/naturalsort/naturalsort-0.1.3-foss-2020b.eb:3]: done

File: easybuild/easyconfigs/c/CRISPR-DAV/CRISPR-DAV-2.3.4-foss-2020b.eb (hunk 8 above)
Comment (deniskristak) [reply easybuild/easyconfigs/c/CRISPR-DAV/CRISPR-DAV-2.3.4-foss-2020b.eb]: correct, renamed

File: easybuild/easyconfigs/p/pysamstats/pysamstats-1.1.2-foss-2020b.eb (hunk 11 above)
Comment (deniskristak) [reply easybuild/easyconfigs/p/pysamstats/pysamstats-1.1.2-foss-2020b.eb:32]: added

File: easybuild/easyconfigs/p/pysamstats/pysamstats-1.1.2-foss-2020b.eb (hunk 4 above)
Comment (deniskristak) [reply easybuild/easyconfigs/p/pysamstats/pysamstats-1.1.2-foss-2020b.eb:19]: done

File: easybuild/easyconfigs/e/Excel-Writer-XLSX/Excel-Writer-XLSX-1.09-foss-2020b.eb (hunk 1 above)
Comment (deniskristak) [reply easybuild/easyconfigs/e/Excel-Writer-XLSX/Excel-Writer-XLSX-1.09-foss-2020b.eb:25]: done

File: easybuild/easyconfigs/f/FLASH/FLASH-2.2.00-foss-2020b.eb (hunk 9 above)
Comment (deniskristak) [reply easybuild/easyconfigs/f/FLASH/FLASH-2.2.00-foss-2020b.eb]: done

File: easybuild/easyconfigs/e/Excel-Writer-XLSX/Excel-Writer-XLSX-1.09-foss-2020b.eb (hunk 12 above)
Comment (deniskristak) [reply easybuild/easyconfigs/e/Excel-Writer-XLSX/Excel-Writer-XLSX-1.09-foss-2020b.eb]: added

**Comment by deniskristak** [comment]
@boegel ready for review

File: easybuild/easyconfigs/p/pysamstats/pysamstats-1.1.2-foss-2020b.eb (hunk 11 above)
Comment (deniskristak) [reply easybuild/easyconfigs/p/pysamstats/pysamstats-1.1.2-foss-2020b.eb:32]: added

File: easybuild/easyconfigs/c/CRISPR-DAV/CRISPR-DAV-2.3.4-foss-2020b.eb (hunk 6 above)
Comment (deniskristak) [reply easybuild/easyconfigs/c/CRISPR-DAV/CRISPR-DAV-2.3.4-foss-2020b.eb]: you answered elsewhere, removing

File: easybuild/easyconfigs/a/ABRA2/ABRA2-2.23-GCC-10.2.0.eb (hunk 7 above)
Comment (boegel) [reply easybuild/easyconfigs/a/ABRA2/ABRA2-2.23-GCC-10.2.0.eb:45]: OK, please add a comment with a reference to that issue then:

```python
# required to work around localization bug, see https://github.com/mozack/abra2/issues/25
modextravars = {'LC_ALL': 'en_US.UTF-8'}
```

File: easybuild/easyconfigs/a/ABRA2/ABRA2-2.23-GCC-10.2.0.eb [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+files_to_copy = [
+    (['abra'], 'bin'),
+    (['target/libAbra.%s' % SHLIB_EXT], 'lib'),
+    'target/abra2-%(version)s-jar-with-dependencies.jar',
+]
+
+postinstallcmds = ["cd %(installdir)s && mv abra2-%(version)s-jar-with-dependencies.jar abra2-%(version)s.jar"]
+
+sanity_check_paths = {
+    'files': ['abra2-%(version)s.jar', 'bin/abra', 'lib/libAbra.%s' % SHLIB_EXT],
```
Comment (boegel) [inline easybuild/easyconfigs/a/ABRA2/ABRA2-2.23-GCC-10.2.0.eb:40]: Can we add a sanity check command for the `abra` command?

File: easybuild/easyconfigs/c/CRISPR-DAV/CRISPR-DAV-2.3.4-foss-2020b.eb [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+sources = ['v%(version)s.tar.gz']
+patches = ['crispr-dav-2.3.4_remove_hardcoding.patch']
+checksums = [
+    '49975cd48bdbf31fe5a9e2aaa3f5ed85d3cc6f65a422ee3aa8daed890159d2ae',  # v2.3.4.tar.gz
+    # crispr-dav-2.3.4-foss2020b_remove_hardcoding.patch
+    '41475a09754d65cc2c88d4161be6eacdfc98242bccbe49778c4f1f74210a6cb2',
+]
+
+# also needs `naturalsort`, which is now an R extension - see https://github.com/easybuilders/easybuild-easyconfigs/pull/13762
+# if missing, try reinstalling R (https://docs.easybuild.io/en/latest/Partial_installations.html#installing-additional-extensions-using-k-skip)
```
Comment (boegel) [inline easybuild/easyconfigs/c/CRISPR-DAV/CRISPR-DAV-2.3.4-foss-2020b.eb:20]: Not needed here, just mention this in a PR description (not the easyconfig file).

There's probably also other R libraries that are required?

If you want to make a note in the easyconfig, do this, that's sufficient:

```python
dependencies = [
    ('R', '4.0.3'),  # provides naturalsort R package

File: easybuild/easyconfigs/c/CRISPR-DAV/CRISPR-DAV-2.3.4-foss-2020b.eb [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+homepage = 'https://github.com/pinetree1/crispr-dav/'
+description = """CRISPR-DAV is a pipeline to analyze
+amplicon-based NGS data of CRISPR clones in a high throughput manner."""
+
+toolchain = {'name': 'foss', 'version': '2020b'}
+
+source_urls = ['https://github.com/pinetree1/crispr-dav/archive/']
+sources = ['v%(version)s.tar.gz']
+patches = ['crispr-dav-2.3.4_remove_hardcoding.patch']
```
Comment (boegel) [inline easybuild/easyconfigs/c/CRISPR-DAV/CRISPR-DAV-2.3.4-foss-2020b.eb:12]: please rename the patch file to `CRISPR-DAV-2.3.4_remove_hardcoding.patch`, and use `%(version)s` here:

```python
patches = ['CRISPR-DAV-%(version)s_remove_hardcoding.patch']

File: easybuild/easyconfigs/p/PRINSEQ/PRINSEQ-0.20.4-foss-2020b-Perl-5.32.0.eb [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+# this is the script which relies in many extra Perl libraries so we verify it's working fine
+sanity_check_commands = ['%(installdir)s/prinseq-graphs.pl']
+
+modextrapaths = {
+    'PATH': '',
+    'PERL5LIB': 'lib/perl5/site_perl/%(perlver)s/',
+}
+
+postinstallcmds = [
+    "sed -i -e 's|/usr/bin/perl|/usr/bin/env\ perl|' %(installdir)s/*.pl",
```
Comment (boegel) [inline easybuild/easyconfigs/p/PRINSEQ/PRINSEQ-0.20.4-foss-2020b-Perl-5.32.0.eb:89]: You should use `fix_perl_shebang_for` for this instead (the path is relative to the install dir):

```python
fix_perl_shebang_for = ["*.pl"]
```

Also, please move `postinstallcmds` (and `fix_perl_shebang_for`) up, above `sanity_check_commands`.

File: easybuild/easyconfigs/p/PRINSEQ/PRINSEQ-0.20.4-foss-2020b-Perl-5.32.0.eb [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+        'source_tmpl': 'Contextual-Return-v%(version)s.tar.gz',
+        'source_urls': ['https://cpan.metacpan.org/authors/id/D/DC/DCONWAY/'],
+        'checksums': ['17a1ed1043a2abe123920894d6023709e834de66ef766be86e4cfbb96b1dd16c'],
+    }),
+]
+
+# this is the script which relies in many extra Perl libraries so we verify it's working fine
+sanity_check_commands = ['%(installdir)s/prinseq-graphs.pl']
+
+modextrapaths = {
```
Comment (boegel) [inline easybuild/easyconfigs/p/PRINSEQ/PRINSEQ-0.20.4-foss-2020b-Perl-5.32.0.eb:83]: please move this down, below `sanity_check_paths`

File: easybuild/easyconfigs/p/PRINSEQ/PRINSEQ-0.20.4-foss-2020b-Perl-5.32.0.eb [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+    "sed -i -e 's|/usr/bin/perl|/usr/bin/env\ perl|' %(installdir)s/*.pl",
+    "chmod +x %(installdir)s/*.pl",
+]
+
+sanity_check_paths = {
+    'files': ['prinseq-lite.pl', 'prinseq-graphs.pl', 'prinseq-graphs-noPCA.pl'],
+    'dirs': []
+}
+
+
```
Comment (boegel) [inline easybuild/easyconfigs/p/PRINSEQ/PRINSEQ-0.20.4-foss-2020b-Perl-5.32.0.eb:98]: remove duplicate empty line

**Review by boegel — CHANGES_REQUESTED** [review]

File: easybuild/easyconfigs/a/ABRA2/ABRA2-2.23-GCC-10.2.0.eb (hunk 13 above)
Comment (boegel) [reply easybuild/easyconfigs/a/ABRA2/ABRA2-2.23-GCC-10.2.0.eb:40]: Looks like the answer is no, it doesn't produce output without options, and doesn't support the usual suspects like `--help`, `-h`, `--version`. Fine.

**Comment by boegel** [comment]
Test report by @boegel
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/2487
**SUCCESS**
Build succeeded for 6 out of 6 (6 easyconfigs in total)
node3585.doduo.os - Linux RHEL 8.2, x86_64, AMD EPYC 7552 48-Core Processor (zen2), Python 3.6.8
See https://gist.github.com/511e361bb35aef80ba46503b62939e36 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/2487
**SUCCESS**
Build succeeded for 6 out of 6 (6 easyconfigs in total)
node3111.skitty.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 3.6.8
See https://gist.github.com/90e614f57b3cca5a4eeb32b9439adba3 for a full test report.

**Comment by boegel** [comment]
@boegelbot please test @ generoso
EB_ARGS="--include-easyblocks-from-pr 2487"

**Review by boegel — APPROVED** [review]

**Comment by boegel** [comment]
Test report by @boegel
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/2487
**SUCCESS**
Build succeeded for 7 out of 7 (6 easyconfigs in total)
node3111.skitty.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 3.6.8
See https://gist.github.com/7adad6dba709de77ae3ddb954ef48a24 for a full test report.

**Review by boegel — APPROVED** [review]
lgtm

**Comment by boegel** [comment]
Going in, thanks @deniskristak!
